src/boot/efi/stub.c

```diff
@@ -180,6 +181,196 @@ static bool use_load_options(
+
+        if (n_items == 0)
+                return EFI_SUCCESS; /* Empty directory */
+
+        /* Now, sort the files we found, to make this uniform and stable (and to ensure the TPM measurements
+         * are not dependent on read order) */
+        sort_pointer_array((void**) items, n_items, (compare_pointer_func_t) strcmp16);
+
+        for (size_t i = 0; i < n_items; i++) {
+                size_t addrs[_UNIFIED_SECTION_MAX] = {}, szs[_UNIFIED_SECTION_MAX] = {};
+                _cleanup_free_ EFI_DEVICE_PATH *addon_path = NULL;
+                _cleanup_(unload_imagep) EFI_HANDLE addon = NULL;
+                EFI_LOADED_IMAGE_PROTOCOL *loaded_addon = NULL;
+                _cleanup_free_ char16_t *addon_spath = NULL;
+
+                addon_spath = xasprintf("%ls\\%ls", prefix, items[i]);
+                err = make_file_device_path(loaded_image->DeviceHandle, addon_spath, &addon_path);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Error making device path for %ls: %m", addon_spath);
+
+                /* By using shim_load_image, we cover both the case where the PE files are signed with MoK
+                 * and with DB, and running with or without shim. */
+                err = shim_load_image(stub_image, addon_path, &addon);
+                if (err != EFI_SUCCESS) {
+                        log_error_status(err,
+                                         "Failed to read '%ls' from '%ls', ignoring: %m",
+                                         items[i],
+                                         addon_spath);
+                        continue;
+                }
+
+                err = BS->HandleProtocol(addon,
+                                         MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL),
+                                         (void **) &loaded_addon);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to find protocol in %ls: %m", items[i]);
+
+                err = pe_memory_locate_sections(loaded_addon->ImageBase, unified_sections, addrs, szs);
+                if (err != EFI_SUCCESS || szs[UNIFIED_SECTION_CMDLINE] == 0) {
+                        if (err == EFI_SUCCESS)
+                                err = EFI_NOT_FOUND;
+                        log_error_status(err,
+                                         "Unable to locate embedded .cmdline section in %ls, ignoring: %m",
+                                         items[i]);
+                        continue;
+                }
+
+                /* We want to enforce that addons are not UKIs, i.e.: they must not embed a kernel. */
+                if (szs[UNIFIED_SECTION_LINUX] > 0) {
+                        log_error_status(EFI_INVALID_PARAMETER, "%ls is a UKI, not an addon, ignoring: %m", items[i]);
+                        continue;
+                }
+
+                /* Also enforce that, in case they are specified, .osrel and .uname match */
```

Comment:
this is a bit weird to me. I think the .osrel match byte-by-byte is a bit weird. We should only require that ID=, VARIANT_ID=, VERSION_ID=, BUILD_ID=, IMAGE_ID=, IMAGE_VERSION=, i.e.

but i guess this kind of relaxing we can do later.

Reply:
The request for this was to do an exact match on a specific UKI, and it works for that. I do not want to invent a new, separate full policy matching, there's sbat for that.

Reply:
"the request"?

I don't understand how sbat would be useful for matching up add-ons with their intended uses? sbat is about revocation, is it not?

Reply:
see earlier comments, someone asked for exact matching too

Reply:
Where? Do you mean this:

https://github.com/systemd/systemd/pull/27358#issuecomment-1554245415

Doesn't suggest exact matching does it?

The thing is: if we compare the human readable strings, why even bother with the non-human-readable strings at all? The idea was to decouple these, so that comparisons are based on computer-readable stuff and not human-readable stuff.

That's exactly how sbat does it too btw: the first two fields in each sbat line are about comparing, the other just decoration for human consumption.

Reply:
It does suggest that, because the request was to be able to do a 1:1 match, ie: you know what UKI you have and want to match the addon to. So it's much easier to do objcopy and extract the .osrel section, and just pass it wholesale to ukify to build the addon, than it is to go and fish a specific parameter. So it doesn't matter if it also matches the readable params. This is just a quick and easy thing, it's not supposed to devolve into a full blown policy, there's sbat for that.

Reply:
Again, i don't see how sbat would work for that. sbat is pretty much an sbom, it doesn't tell us anything about compatibility.

And the fact that you can copy out/reuse osrel is fine, that does not imply whatsoever that this is how we do comparisons, too.

Reply:
It's not about compatibility, it's about revocation. Deploying a new sbat policy an entire generation of addons can be revoked, whether it's 1, 10 or 10000.

Reply:
yeah, i understand sbat I think. But I am not talking about sbat here, you are. I was talking about matching osrel against each other, and I think we should do that according to the semantics we defined, and not byte-by-byte because that's simply not how we generally detect compat for osrel.

In fact thinking more abou this I think we probably should go further and compare addons compat like we check sysext + confext compat, i.e. also explicitly check some "LEVEL" field or so.

Reply:
I've just dropped .osrel and instead left .uname, can be used for the exact same purpose anyway